Fix channel frame handling upon framing error

vitaly-krugl opened the pull request:
This PR, together with PR #73, should fully address issue #72

Drop non-close/close_ok methods per AMQP stability rule when closing due to input framing error.

Fixed up/improved a number of tests, inluding two with duplicate names.

Don't override system-level exceptions (e.g., SystemExit, KeyboardException) in Channel.process_frames()

vitaly-krugl commented:
@awestendorf: this PR, together with PR #73 should fully address issue #72

vitaly-krugl commented:
Per http://www.rabbitmq.com/resources/specs/amqp0-9-1.extended.xml, re. channel.close:

> After sending this method, any received methods except Close and Close-OK MUST be discarded. ...

awestendorf commented on tests/unit/channel_test.py:
```diff
@@ -237,7 +237,7 @@ def test_buffer_frame(self):
         # Not that this should ever happen, but to be sure
         c = Channel(None, None, {})
-        stub(c.dispatch)
+        expect(c.dispatch).times(0)
```
You don't need this, that's essentially what `stub` is doing. Do you prefer this usage?

vitaly-krugl replied:
I wasn't aware of that initially, so I thought that it wasn't testing anything; sort of like patching something in the mock library, but then not checking call count.

However, once I learned the semantics of stub better, I realized that I still liked using `times(0)` in such cases, because it's explicit vs. implicit. I'd be happy to change that back. Please let me know.

awestendorf replied:
You're welcome to whatever you'd like. I don't get much feedback on Chai usage, so it's good to know your perspective. I think I'll add a note to the documentation to call out how those are equivalent.

awestendorf commented:
I see what you're doing here and I like it, but I want to be sure that this doesn't break anything, so it will take some testing and review before I'm comfortable merging.

vitaly-krugl commented:
@awestendorf, thanks for reviewing. This one is far less urgent for me than #73.

This basically fixes the secondary issues that I discovered and described when investigating issue #72:
1. The channel was getting confused by incoming framing error and was unable to recover (internal exceptions resulting in recursive Channel.close calls,  and finally hang)
2. It was overriding system-exiting exceptions, preventing things such as SystemExit and KeyboardException from working properly.

In this PR, I took the cautious approach of only implementing the stability rule when closing the channel under the above-described framing error duress conditions.

I limited the scope to this change to the framing error scenario out of concern for creating an unintended side-effect under normal Channel.close conditions. 

For example, in the basic.return failure scenario, the basic.return frames were getting chewed up and causing internal exceptions and Channel.close recursion, and if we didn't start dropping non-close-related frames, then the user code listening for Basic.Ack, would have its "ack listener" callback called, and would have mistakenly concluded that the message was delivered successfully and marked the operation as successfully-completed, which would be a mistake.

vitaly-krugl commented on tests/unit/channel_test.py:
```diff
@@ -274,7 +274,7 @@ def test_process_frames_when_connectionclosed_on_dispatch(self):
         expect(c.dispatch).args(f0).raises(
             ConnectionClosed('something darkside'))
-        stub(c.close)  # assert not called
+        expect(c.close).times(0)
```
@awestendorf, here is a better example: there used to be a stub with the comment here `stub(c.close)  # assert not called`, so I am thinking that instead of explaining what the code does, it might be self-documenting to simply use the mechanism that makes it explicit: `expect(c.close).times(0)` (does what it says, says what it does)

awestendorf commented on haigha/channel.py:
```diff
@@ -227,10 +235,25 @@ def process_frames(self):
+                if (not isinstance(frame, MethodFrame) or
+                      frame.class_id != 20 or
+                      frame.method_id not in (40, 41)):
+                    self.logger.warn("Emergency channel close: dropping input "
```
I don't like that this breaks the abstraction layer, wherein the class id and method id is defined between the `Connection` and the individual `ProtocolClass` es. I can see where that could be challenging to resolve.

vitaly-krugl replied:
Aaron, what do you recommend? Should I add a couple of methods in BasicClass along the lines of: `is_close_method(method_frame)` and `is_close_ok_method(method_frame)` in order to protect the abstraction?

awestendorf replied:
I'm not sure yet

vitaly-krugl commented:
Hi Aaron, we've been running with this change in production error-free 24/7 for several months now. Prior to this PR, Haigha would hang when encountering an unexpected frame as discovered with missing `Basic.Return` support and documented in issue #72. It was also overriding system-level exceptions (e.g., SystemExit, KeyboardException) in Channel.process_frames().

vitaly-krugl commented:
rebased against current master

vitaly-krugl commented on haigha/classes/channel_class.py:
```diff
@@ -15,6 +15,12 @@ class ChannelClass(ProtocolClass):
+
+    # Channel method ids for error-recovery code in Channel
+    CLOSE_METHOD_ID = 40
+    CLOSE_OK_METHOD_ID = 41
```
Could define all of the 5 channel method ids here, and then use them instead of the "naked" constants in `self. dispatch_map` initialization.

vitaly-krugl commented:
@awestendorf: Hi Aaron, I addressed your outstanding concern about the use of the "naked" constants for class and method identifiers via 9ec77191667b65b7e65483ce18950022c4e53469 in this PR and rebased against the current master.

awestendorf commented:
@vitaly-krugl I forgot about this PR. I'd like to find a way to abstract this, so that the `ChannelClass` ids don't bleed into the `Channel` abstraction. I may just have to get over it and consider an alternative relationship, like I did with `Connection` and `ConnectionChannel`.

Since I've sat on this so long and my only hangups are with the abstraction, I'm going to merge it any way. It's more important for it to be technically correct than to wait for a potential improvement to the code structure. I can always revisit that at a later date.
